# Análisis inicial y trata de datos

In [1]:
# set PYTHONPATH="${PYTHONPATH}:/path/to/your/project/"

In [2]:
# Importamos las librerías necesarias
import pandas as pd # type: ignore
import numpy as np # type: ignore
# from src import funciones_soporte as fs
import psycopg2 as ps # type: ignore

In [3]:
# Importamos el archivo con el que vamos a trabajar, que en este caso es un parquet. Para que funcione el read_parquet es necesario pasarle
# el parámetro ' engine = auto/pyarrow', ambos funcionan. En caso de que se use pyarrow será necesario realizar su instalaciñon previamnente en el entorno.
df_raw = pd.read_parquet("../data/reservas_hoteles.parquet", engine='auto')
df_raw.head()

id_reserva                            id_cliente  \
0  5256cc90-139b-43d2-8ec5-412495d751cf  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
1  84fd6209-bd8d-4a92-bfe6-c68ee2c49271  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
2  40c4cb55-d1f5-407b-832f-4756b8ff77b4  6ec4d7d2-e382-47cc-94f6-c647577d4035   
3  f2ce8df5-7844-43e1-8c0f-97ce7a208f21  aed17fe8-eb18-4f69-bf08-df1afdc339c3   
4  57d4515a-447d-4067-afcb-9bdcf4e4e915  710a2fbb-264f-426a-aa66-f8fffd146a2d   

      nombre   apellido                          mail  competencia  \
0  Feliciana     Cantón  feliciana.cantón@example.com         True   
1   Leonardo     Hierro   leonardo.hierro@example.com         True   
2      Maite  Calatayud   maite.calatayud@example.com        False   
3      Tecla      Bonet       tecla.bonet@example.com        False   
4    Amílcar     Andrés    amílcar.andrés@example.com        False   

  fecha_reserva inicio_estancia final_estancia  id_hotel  precio_noche  \
0                    2025-03-01     2025-03-02       113           NaN   
1                    2025-03-01     2025-03-02       194           NaN   
2    2025-02-09      2025-03-01     2025-03-02        14        119.96   
3    2025-02-08      2025-03-01     2025-03-02        38        211.53   
4    2025-02-02      2025-03-01     2025-03-02        35        378.37   

           nombre_hotel  estrellas  ciudad  
0                              NaN          
1                              NaN          
2     Hotel Monte Verde        1.0  Madrid  
3  Hotel Brisas del Mar        5.0  Madrid  
4  Hotel Camino del Sol        1.0  Madrid

In [4]:
# Creamos una copia de los datos originales
data = df_raw.copy()
data.head()

id_reserva                            id_cliente  \
0  5256cc90-139b-43d2-8ec5-412495d751cf  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
1  84fd6209-bd8d-4a92-bfe6-c68ee2c49271  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
2  40c4cb55-d1f5-407b-832f-4756b8ff77b4  6ec4d7d2-e382-47cc-94f6-c647577d4035   
3  f2ce8df5-7844-43e1-8c0f-97ce7a208f21  aed17fe8-eb18-4f69-bf08-df1afdc339c3   
4  57d4515a-447d-4067-afcb-9bdcf4e4e915  710a2fbb-264f-426a-aa66-f8fffd146a2d   

      nombre   apellido                          mail  competencia  \
0  Feliciana     Cantón  feliciana.cantón@example.com         True   
1   Leonardo     Hierro   leonardo.hierro@example.com         True   
2      Maite  Calatayud   maite.calatayud@example.com        False   
3      Tecla      Bonet       tecla.bonet@example.com        False   
4    Amílcar     Andrés    amílcar.andrés@example.com        False   

  fecha_reserva inicio_estancia final_estancia  id_hotel  precio_noche  \
0                    2025-03-01     2025-03-02       113           NaN   
1                    2025-03-01     2025-03-02       194           NaN   
2    2025-02-09      2025-03-01     2025-03-02        14        119.96   
3    2025-02-08      2025-03-01     2025-03-02        38        211.53   
4    2025-02-02      2025-03-01     2025-03-02        35        378.37   

           nombre_hotel  estrellas  ciudad  
0                              NaN          
1                              NaN          
2     Hotel Monte Verde        1.0  Madrid  
3  Hotel Brisas del Mar        5.0  Madrid  
4  Hotel Camino del Sol        1.0  Madrid

In [5]:
# función info
def info_df(dataframe):
    """
    Función que devuelve información general sobre el DatFrame que le pasemos.


    Args:
        df (DataFrame): DataFrame con información que queramos revisar

    Returns:
        DataFrame: DataFrame con información general sobre las columnas del DataFrame que se le ha pasado a la función: tipo de datos, número de
        registros, número de valores nulos, porcentaje de los valores nulos sobre el total
    """
    info_df = pd.DataFrame()
    info_df["Tipo_dato"] = dataframe.dtypes
    info_df["numero_registros"] = [dataframe[elemento].value_counts().sum() for elemento in dataframe]
    info_df["Numero_nulos"] = round(dataframe.isnull().sum())
    info_df["%_nulos"] = round((dataframe.isnull().sum()/dataframe.shape[0])*100, 2)

    return info_df

In [6]:
# Información general del dataframe
info_df(data)

Tipo_dato  numero_registros  Numero_nulos  %_nulos
id_reserva         object             15098             0     0.00
id_cliente         object             15098             0     0.00
nombre             object             15098             0     0.00
apellido           object             15098             0     0.00
mail               object             15098             0     0.00
competencia          bool             15098             0     0.00
fecha_reserva      object             15098             0     0.00
inicio_estancia    object             15023            75     0.50
final_estancia     object             15023            75     0.50
id_hotel            int64             15098             0     0.00
precio_noche      float64              9874          5224    34.60
nombre_hotel       object             15098             0     0.00
estrellas         float64              9926          5172    34.26
ciudad             object             15098             0     0.00

In [7]:
# Vamos a ver las columnas que tenemos en el dataframe
data.columns

Index(['id_reserva', 'id_cliente', 'nombre', 'apellido', 'mail', 'competencia',
       'fecha_reserva', 'inicio_estancia', 'final_estancia', 'id_hotel',
       'precio_noche', 'nombre_hotel', 'estrellas', 'ciudad'],
      dtype='object')

Tras realizar una overview de los datos recibidos observamos las siguientes casuísticas:
- Tenemos 14 columnas.
- Hay 3 columnas con información sobre fechas, las cuales son de tipo object. Será necesario modificar el tipo de dato de las mismas para poder trabajar más adelante.
- La columna competencia es un booleano, el cual habrá que ver con que se corresponde true y false.
- Tenemos tres columnas de id: de reserva, de cliente y de hotel. Las dos primeras con tipo object, y la última con tipo int. Cambiaremos la de id_hotel a tipo object para tenerlas las tres iguales. 
- Hay existencia de nulos en cuatro columnas:
    
    **a)** inicio_estancia: tenemos un 0,5% de nulos.
    
    **b)** fin_estancia: tenemos un 0,5% de nulos. Comprobaremos si son los mismos que la columna de inicio estancia, revisar casuística y tratar.
    
    **c)** precio_noche: hay un 34,6% de nulos, un valor significativo. Será necesario tratarlos.
    
    **d)** estrellas: hay un 34,26% de nulos, un valor significativo. Será necesario tratarlos. Asimismo el tipo de dato de esta columna es float. Podemos plantear modificarlo a tipo object, a un str. 

A continuación realizamos las modificaciones necesarias de las columnas para poder trabajar con ellas. 

## Búsqueda de duplicados completamente iguales 

In [43]:
# pd.set_option('display.max_rows', None)

In [72]:
# Vamos a revisar la existencia de duplicados completamente iguales en el df. Revisaremos si se pueden eliminar
data.duplicated().sum()
# Tenemos 98 elementos duplicados en total en el dataframe
data = data.drop_duplicates()

## Modificación columnas de fecha

In [73]:
# Creamos una función que nos modifica ek tipo de dato de las columnas que le pasemos a tipo fecha
def data_fechas(dataframe, columnas):
    for col in columnas:
        dataframe[col] = pd.to_datetime(dataframe[col])
    return dataframe

In [74]:
# modificamos las columnas que queremos a tipo fecha
data = data_fechas(data, ["fecha_reserva", "inicio_estancia", "final_estancia"])

In [75]:
# comprobamos que la modificación se ha realizado de forma correcta
info_df(data)

Tipo_dato  numero_registros  Numero_nulos  %_nulos
id_reserva               object             15000             0     0.00
id_cliente               object             15000             0     0.00
nombre                   object             15000             0     0.00
apellido                 object             15000             0     0.00
mail                     object             15000             0     0.00
competencia                bool             15000             0     0.00
fecha_reserva    datetime64[ns]              9828          5172    34.48
inicio_estancia  datetime64[ns]             14925            75     0.50
final_estancia   datetime64[ns]             14925            75     0.50
id_hotel                 object             15000             0     0.00
precio_noche            float64              9776          5224    34.83
nombre_hotel             object             15000             0     0.00
estrellas               float64              9828          5172    34.48
ciudad                   object             15000             0     0.00
col_prueba               object             15000             0     0.00

Modificación realizada con éxito.

## Columna valores booleanos

In [76]:
data.head()

id_reserva                            id_cliente  \
0  5256cc90-139b-43d2-8ec5-412495d751cf  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
1  84fd6209-bd8d-4a92-bfe6-c68ee2c49271  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
2  40c4cb55-d1f5-407b-832f-4756b8ff77b4  6ec4d7d2-e382-47cc-94f6-c647577d4035   
3  f2ce8df5-7844-43e1-8c0f-97ce7a208f21  aed17fe8-eb18-4f69-bf08-df1afdc339c3   
4  57d4515a-447d-4067-afcb-9bdcf4e4e915  710a2fbb-264f-426a-aa66-f8fffd146a2d   

      nombre   apellido                          mail  competencia  \
0  Feliciana     Cantón  feliciana.cantón@example.com         True   
1   Leonardo     Hierro   leonardo.hierro@example.com         True   
2      Maite  Calatayud   maite.calatayud@example.com        False   
3      Tecla      Bonet       tecla.bonet@example.com        False   
4    Amílcar     Andrés    amílcar.andrés@example.com        False   

  fecha_reserva inicio_estancia final_estancia id_hotel  precio_noche  \
0           NaT      2025-03-01     2025-03-02      113           NaN   
1           NaT      2025-03-01     2025-03-02      194           NaN   
2    2025-02-09      2025-03-01     2025-03-02       14        119.96   
3    2025-02-08      2025-03-01     2025-03-02       38        211.53   
4    2025-02-02      2025-03-01     2025-03-02       35        378.37   

           nombre_hotel  estrellas  ciudad col_prueba  
0                              NaN               True  
1                              NaN               True  
2     Hotel Monte Verde        1.0  Madrid      False  
3  Hotel Brisas del Mar        5.0  Madrid      False  
4  Hotel Camino del Sol        1.0  Madrid      False

In [77]:
# Al intentar filtrar la columna en sí, como es un booleano no nos devuelve nada, por lo que creamos una copia de la columna de tipo str
# a través de la cual podremos realizar el filtrado entre true y false.
data["col_prueba"] = data["competencia"].astype(str)
data.head()

id_reserva                            id_cliente  \
0  5256cc90-139b-43d2-8ec5-412495d751cf  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
1  84fd6209-bd8d-4a92-bfe6-c68ee2c49271  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
2  40c4cb55-d1f5-407b-832f-4756b8ff77b4  6ec4d7d2-e382-47cc-94f6-c647577d4035   
3  f2ce8df5-7844-43e1-8c0f-97ce7a208f21  aed17fe8-eb18-4f69-bf08-df1afdc339c3   
4  57d4515a-447d-4067-afcb-9bdcf4e4e915  710a2fbb-264f-426a-aa66-f8fffd146a2d   

      nombre   apellido                          mail  competencia  \
0  Feliciana     Cantón  feliciana.cantón@example.com         True   
1   Leonardo     Hierro   leonardo.hierro@example.com         True   
2      Maite  Calatayud   maite.calatayud@example.com        False   
3      Tecla      Bonet       tecla.bonet@example.com        False   
4    Amílcar     Andrés    amílcar.andrés@example.com        False   

  fecha_reserva inicio_estancia final_estancia id_hotel  precio_noche  \
0           NaT      2025-03-01     2025-03-02      113           NaN   
1           NaT      2025-03-01     2025-03-02      194           NaN   
2    2025-02-09      2025-03-01     2025-03-02       14        119.96   
3    2025-02-08      2025-03-01     2025-03-02       38        211.53   
4    2025-02-02      2025-03-01     2025-03-02       35        378.37   

           nombre_hotel  estrellas  ciudad col_prueba  
0                              NaN               True  
1                              NaN               True  
2     Hotel Monte Verde        1.0  Madrid      False  
3  Hotel Brisas del Mar        5.0  Madrid      False  
4  Hotel Camino del Sol        1.0  Madrid      False

`Hoteles de la competencia (col True)`

Estas columnas entendemos que se consideran como competencia del grupo, por lo que nos falta información de las mismas.

Tenemos valores nulos en este caso en las columnas de inicio y fin de estancia (23 en cada una), los cuales se corresponden en ambas columnas con los mismos clientes. Asimismo hay valores nulos en la columna de precio de noche (5172), que entendemos que se corresponde con los valores nulos de los hoteles de los cuales no tenemos estrellas (nulos en columna estrellas). Existen valores nulos en la columna de fecha de reserva, de la cual no tenemos ninguna información al respecto. 

Por otro lado no tenemos los nombres de estos hoteles, sino que están puestos en el dataframe como espacios en blanco.

Para tratar estos datos realizaremos un scrapeo de los nombres de los hoteles correspondientes, sus estrellas y el precio por noche de cada uno. Una vez hayamos obtenido la información necesaria, les asignaremos un id de hotel para poder trabajar en el dataframe.


Tenemos valores duplicados en la columna de id de cliente, será necesario eliminar estos valores. 

De estos hoteles todas las estancias se han realizado entre el 01/03/2025 y el 02/03/2025. Podemos considerar rellenar el resto de fechas con los mismos valores.

In [78]:
# Filtramos por los valores que son True
hoteles_competencia = data[data["col_prueba"] == "True"]
hoteles_competencia.head()

id_reserva  \
0   5256cc90-139b-43d2-8ec5-412495d751cf   
1   84fd6209-bd8d-4a92-bfe6-c68ee2c49271   
7   4d65096e-5f12-4ce7-9fd6-dd20e3cf5194   
9   458029ad-cc9b-4d85-8290-fec170bcc02d   
11  ac0fe86e-90a0-4fe6-a02e-6ea00e9a8621   

                              id_cliente        nombre apellido  \
0   346f2a77-33f4-4b2d-82f3-e8501e21b28d     Feliciana   Cantón   
1   346f2a77-33f4-4b2d-82f3-e8501e21b28d      Leonardo   Hierro   
7   443a7307-2183-4051-b0bb-26f12fc1f478        Judith    Rivas   
9   e1a34338-a0aa-4b0e-96de-55f884eeb074        Adolfo   Cuervo   
11  2949fb90-39ab-4613-b4c1-e238bb56faf3  María Carmen  Navarro   

                                mail  competencia fecha_reserva  \
0       feliciana.cantón@example.com         True           NaT   
1        leonardo.hierro@example.com         True           NaT   
7           judith.rivas@example.com         True           NaT   
9          adolfo.cuervo@example.com         True           NaT   
11  maría carmen.navarro@example.com         True           NaT   

   inicio_estancia final_estancia id_hotel  precio_noche nombre_hotel  \
0       2025-03-01     2025-03-02      113           NaN                
1       2025-03-01     2025-03-02      194           NaN                
7       2025-03-01     2025-03-02      131           NaN                
9       2025-03-01     2025-03-02      114           NaN                
11      2025-03-01     2025-03-02      103           NaN                

    estrellas ciudad col_prueba  
0         NaN              True  
1         NaN              True  
7         NaN              True  
9         NaN              True  
11        NaN              True

In [48]:
info_df(hoteles_competencia)

Tipo_dato  numero_registros  Numero_nulos  %_nulos
id_reserva               object              5172             0     0.00
id_cliente               object              5172             0     0.00
nombre                   object              5172             0     0.00
apellido                 object              5172             0     0.00
mail                     object              5172             0     0.00
competencia                bool              5172             0     0.00
fecha_reserva    datetime64[ns]                 0          5172   100.00
inicio_estancia  datetime64[ns]              5149            23     0.44
final_estancia   datetime64[ns]              5149            23     0.44
id_hotel                 object              5172             0     0.00
precio_noche            float64                 0          5172   100.00
nombre_hotel             object              5172             0     0.00
estrellas               float64                 0          5172   100.00
ciudad                   object              5172             0     0.00
col_prueba               object              5172             0     0.00

In [79]:
# Comprobamos si los valores nulos de las columnas de inicio y final de estancia se corresponden con los mismos clientes en ambas columnas.
# hoteles_competencia[hoteles_competencia["inicio_estancia"].isnull()] == hoteles_competencia[hoteles_competencia["final_estancia"].isnull()] 

In [59]:
hoteles_competencia["fecha_reserva"].unique()

<DatetimeArray>
['NaT']
Length: 1, dtype: datetime64[ns]

In [80]:
hoteles_competencia["inicio_estancia"].unique()

<DatetimeArray>
['2025-03-01 00:00:00', 'NaT']
Length: 2, dtype: datetime64[ns]

In [81]:
hoteles_competencia["final_estancia"].unique()

<DatetimeArray>
['2025-03-02 00:00:00', 'NaT']
Length: 2, dtype: datetime64[ns]

In [82]:
# Comprobamos los nombres de los hoteles de la competencia
hoteles_competencia["nombre_hotel"].unique()

array([''], dtype=object)

In [83]:
# Comprobamos los id de hotel de la competencia
hoteles_competencia["id_hotel"].unique()

array(['113', '194', '131', '114', '103', '181', '128', '186', '135',
       '117'], dtype=object)

In [84]:
hoteles_competencia.duplicated().sum()

np.int64(0)

In [85]:
# Revisamos duplicados
hoteles_competencia[hoteles_competencia["id_cliente"].duplicated()]

id_reserva  \
1      84fd6209-bd8d-4a92-bfe6-c68ee2c49271   
1333   35888670-8393-4a86-b10f-42d8cc0cc558   
2467   20c86be7-6537-4d1e-8553-3e95b49af0b9   
3777   beda58ca-7f20-488c-b680-b6d748f666c5   
3944   3e9c3923-4b34-4566-ab96-74237b4f4eae   
4826   fbba658c-f025-484c-a3a4-7a7a8af42863   
8373   8b3a1b34-68dd-4429-93fa-c561ca8137c6   
8551   dcfaed8a-ce16-47f6-8eae-69c410a4ea8b   
11055  9dcb96fc-ac04-4d47-b7e3-b19705c0b5bd   
11858  9387407f-a17f-4ccb-b0b9-fc8e77b961ed   
14625  71bfa845-b10b-42b3-9e09-fcb238341936   

                                 id_cliente            nombre    apellido  \
1      346f2a77-33f4-4b2d-82f3-e8501e21b28d          Leonardo      Hierro   
1333   bfbb9754-fcab-4580-bb2a-9ab295149227           Teodora        Seco   
2467   37ec97b1-9ceb-48e0-a3af-7a79b9585603               Ruy        Ríos   
3777   2b15a636-76f1-4c9a-b689-0581c3245aa2               Ana  Valenzuela   
3944   a44bb598-0a86-4e4a-b19e-66ddff9f3d2f           Manuela     Fonseca   
4826   21294898-0afc-489d-8f71-1ca78bedbe2f              Lina       Ariño   
8373   5a2d9076-03e5-482c-a7a4-45f5abc6cc62            Álvaro     Pascual   
8551   f3e84e3b-35a9-4207-a6a1-4cff7487f569          Consuelo     Estévez   
11055  f8fc8162-f481-44f2-ac98-4a952cbe8ffc  Francisco Javier       Sainz   
11858  9e659fbd-bcc6-4236-9787-6bca04d567c1           Ciríaco       Viana   
14625  1ae69965-f7b0-46d1-ab3a-0b6d207db369              Chus     Uriarte   

                                     mail  competencia fecha_reserva  \
1             leonardo.hierro@example.com         True           NaT   
1333             teodora.seco@example.com         True           NaT   
2467                 ruy.ríos@example.com         True           NaT   
3777           ana.valenzuela@example.com         True           NaT   
3944          manuela.fonseca@example.com         True           NaT   
4826               lina.ariño@example.com         True           NaT   
8373           álvaro.pascual@example.com         True           NaT   
8551         consuelo.estévez@example.com         True           NaT   
11055  francisco javier.sainz@example.com         True           NaT   
11858           ciríaco.viana@example.com         True           NaT   
14625            chus.uriarte@example.com         True           NaT   

      inicio_estancia final_estancia id_hotel  precio_noche nombre_hotel  \
1          2025-03-01     2025-03-02      194           NaN                
1333       2025-03-01     2025-03-02      128           NaN                
2467       2025-03-01     2025-03-02      131           NaN                
3777       2025-03-01     2025-03-02      103           NaN                
3944       2025-03-01     2025-03-02      117           NaN                
4826       2025-03-01     2025-03-02      194           NaN                
8373       2025-03-01     2025-03-02      103           NaN                
8551       2025-03-01     2025-03-02      131           NaN                
11055      2025-03-01     2025-03-02      114           NaN                
11858      2025-03-01     2025-03-02      181           NaN                
14625      2025-03-01     2025-03-02      181           NaN                

       estrellas ciudad col_prueba  
1            NaN              True  
1333         NaN              True  
2467         NaN              True  
3777         NaN              True  
3944         NaN              True  
4826         NaN              True  
8373         NaN              True  
8551         NaN              True  
11055        NaN              True  
11858        NaN              True  
14625        NaN              True

In [95]:
# hoteles_competencia

`Hoteles del grupo (col False)`

Estas columnas entendemos que no se consideran como competencia del grupo, por lo que tenemos casi toda la información de las mismas. Tenemos los nombres de los hoteles del grupo, junto a sus id correspondientes.

Tenemos valores nulos en este caso en las columnas de inicio, fin de estancia y precio de noche (52 en cada una), los cuales se corresponden en todas las columnas con los mismos clientes.

Tenemos duplicados en las columnas de id de cliente (178) e id de reserva (98)

In [87]:
# Filtramos por las columnas que contienen False.
hoteles_grupo = data[data["col_prueba"] == "False"]
hoteles_grupo.head()

id_reserva                            id_cliente  \
2  40c4cb55-d1f5-407b-832f-4756b8ff77b4  6ec4d7d2-e382-47cc-94f6-c647577d4035   
3  f2ce8df5-7844-43e1-8c0f-97ce7a208f21  aed17fe8-eb18-4f69-bf08-df1afdc339c3   
4  57d4515a-447d-4067-afcb-9bdcf4e4e915  710a2fbb-264f-426a-aa66-f8fffd146a2d   
5  ccbd9fe9-5a60-4f6f-bed3-73a7158ecba7  df42ae3d-0b9b-4c06-b22e-939323833ef8   
6  99c3dc4f-663c-45f7-849a-ac9313f3746a  a16a9d83-6888-4947-a7ab-ed41203e347c   

    nombre   apellido                         mail  competencia fecha_reserva  \
2    Maite  Calatayud  maite.calatayud@example.com        False    2025-02-09   
3    Tecla      Bonet      tecla.bonet@example.com        False    2025-02-08   
4  Amílcar     Andrés   amílcar.andrés@example.com        False    2025-02-02   
5     Joan    Vazquez     joan.vazquez@example.com        False    2025-02-08   
6    Chelo       Flor       chelo.flor@example.com        False    2025-02-09   

  inicio_estancia final_estancia id_hotel  precio_noche  \
2      2025-03-01     2025-03-02       14        119.96   
3      2025-03-01     2025-03-02       38        211.53   
4      2025-03-01     2025-03-02       35        378.37   
5      2025-03-01     2025-03-02       36         85.45   
6      2025-03-01     2025-03-02       43        309.80   

             nombre_hotel  estrellas  ciudad col_prueba  
2       Hotel Monte Verde        1.0  Madrid      False  
3    Hotel Brisas del Mar        5.0  Madrid      False  
4    Hotel Camino del Sol        1.0  Madrid      False  
5  Hotel Puerta del Cielo        5.0  Madrid      False  
6      Hotel Encanto Real        1.0  Madrid      False

In [88]:
info_df(hoteles_grupo)

Tipo_dato  numero_registros  Numero_nulos  %_nulos
id_reserva               object              9828             0     0.00
id_cliente               object              9828             0     0.00
nombre                   object              9828             0     0.00
apellido                 object              9828             0     0.00
mail                     object              9828             0     0.00
competencia                bool              9828             0     0.00
fecha_reserva    datetime64[ns]              9828             0     0.00
inicio_estancia  datetime64[ns]              9776            52     0.53
final_estancia   datetime64[ns]              9776            52     0.53
id_hotel                 object              9828             0     0.00
precio_noche            float64              9776            52     0.53
nombre_hotel             object              9828             0     0.00
estrellas               float64              9828             0     0.00
ciudad                   object              9828             0     0.00
col_prueba               object              9828             0     0.00

In [24]:
# Comprobamos si los valores nulos de las columnas de inicio y final de estancia se corresponden con los mismos clientes en ambas columnas.
# df_false[df_false["inicio_estancia"].isnull()] == df_false[df_false["final_estancia"].isnull()]

In [98]:
hoteles_grupo["inicio_estancia"].unique()

<DatetimeArray>
['2025-03-01 00:00:00', 'NaT']
Length: 2, dtype: datetime64[ns]

In [96]:
hoteles_grupo["final_estancia"].unique()

<DatetimeArray>
['2025-03-02 00:00:00', 'NaT']
Length: 2, dtype: datetime64[ns]

In [89]:
# Vamos a revisar los nombres de los hoteles y su id_correspondiente
hoteles_grupo[["id_hotel","nombre_hotel"]]

id_hotel            nombre_hotel
2           14       Hotel Monte Verde
3           38    Hotel Brisas del Mar
4           35    Hotel Camino del Sol
5           36  Hotel Puerta del Cielo
6           43      Hotel Encanto Real
8            6         Palacio del Sol
10           3  Hotel Jardines del Rey
13          29     Hotel Las Estrellas
15          47       Gran Hotel Madrid
17          14      Hotel Torre Dorada
18          27  Hotel Palacio Imperial
19          47    Hotel Camino del Sol
20          13     Hotel Luz de Madrid
21          28      Hotel Encanto Real
22          28     Hotel Los Almendros
23          36  Hotel Palacio Imperial
24          29        Hotel Sol y Luna
25          29       Hotel Monte Verde
26           3      Hotel Mirador Real
27          40  Hotel Puerta del Cielo
28          28     Hotel Las Estrellas
29          40       Hotel Monte Verde
30          43  Hotel Puerta del Cielo
32           6     Hotel Rincón Sereno
33          29      Hotel Vista Alegre
34           3     Hotel Los Almendros
36           2     Hotel Luz de Madrid
37          40        Hotel Costa Azul
38          48        Hotel Costa Azul
39           6       Gran Hotel Madrid
40          35      Hotel Encanto Real
42          29      Hotel Torre Dorada
44          13         Palacio del Sol
45          13       Gran Hotel Madrid
48           2  Hotel Palacio Imperial
49          14      Hotel Encanto Real
51          35    Hotel Camino del Sol
53           3     Hotel Rincón Sereno
55          36     Hotel Luz de Madrid
57          13    Hotel Camino del Sol
58          15       Gran Hotel Madrid
59          35       Gran Hotel Madrid
60          40    Hotel Maravilla Real
61          40     Hotel Las Estrellas
63          36      Hotel Torre Dorada
64          35     Hotel Los Almendros
65          33  Hotel Jardines del Rey
66          36  Hotel Puerta del Cielo
68          48      Hotel Torre Dorada
69           2      Hotel Vista Alegre
71          15    Hotel Camino del Sol
73          48    Hotel Camino del Sol
74          38      Hotel Vista Alegre
75          38      Hotel Vista Alegre
76           2  Hotel Jardines del Rey
77          14       Gran Hotel Madrid
79          48      Hotel Vista Alegre
80          49     Hotel Los Almendros
82          28    Hotel Brisas del Mar
83          13  Hotel Puerta del Cielo
85          35  Hotel Jardines del Rey
86          43     Hotel Los Almendros
87          47     Hotel Las Estrellas
92          14         Palacio del Sol
94           2  Hotel Palacio Imperial
96          47  Hotel Jardines del Rey
101         29        Hotel Sol y Luna
102         14      Hotel Mirador Real
103         13       Hotel Monte Verde
104         40      Hotel Torre Dorada
106         40    Hotel Camino del Sol
107          6       Hotel Monte Verde
110         15       Hotel Monte Verde
111         29     Hotel Los Almendros
113         19    Hotel Camino del Sol
114          2      Hotel Torre Dorada
115         49  Hotel Jardines del Rey
118          6       Hotel Monte Verde
120         15  Hotel Jardines del Rey
121         47      Hotel Encanto Real
124         27      Hotel Vista Alegre
126         38        Hotel Costa Azul
128         35         Palacio del Sol
129         49    Hotel Camino del Sol
134         49     Hotel Rincón Sereno
135         47      Hotel Mirador Real
136         47      Hotel Encanto Real
137         33    Hotel Camino del Sol
138         47      Hotel Vista Alegre
139         13  Hotel Palacio Imperial
144         36    Hotel Brisas del Mar
149         47      Hotel Mirador Real
150         36     Hotel Luz de Madrid
151         28    Hotel Camino del Sol
152         19        Hotel Sol y Luna
153         28     Hotel Las Estrellas
154         33       Hotel Monte Verde
155         36     Hotel Los Almendros
156         27    Hotel Camino del Sol
157         43      Hotel Encanto Real
158         47        Hotel Sol y Luna
159         28         Palac

In [90]:
hoteles_grupo.duplicated().sum()

np.int64(0)

In [91]:
# Revisamos duplicados de los valores 
hoteles_grupo["id_reserva"].duplicated().sum()

np.int64(0)

In [ ]:
hoteles_grupo["id_cliente"].duplicated().sum()

np.int64(80)

In [ ]:
hoteles_grupo[hoteles_grupo["id_cliente"].duplicated()]

id_reserva  \
557    35c7c9d8-4898-4d05-9c34-099e71348a64   
583    172dc5c0-ed32-4f1c-ac40-21f25d3e0742   
592    d75f9705-4c2a-4f9f-a080-584a94eb72b5   
653    47b96396-8992-40d5-a15f-0ad4c74ca2c2   
686    5b3f7171-ea11-4cd5-9e8f-037085d190c0   
837    5c3855a1-981c-4deb-a9e1-7ec90bd337ae   
1134   66076c28-3a21-4e4d-a115-63fd30027a4f   
1232   2be5f54f-7531-4137-9a06-80fb65ac0dac   
1279   e558fc57-2ccc-4610-9c81-098a7fc614a1   
1331   56239530-6705-47bd-8d8c-6942cf2b06e9   
1751   60e07469-1c7b-457c-875e-5d5c32706345   
1866   74ef8f7e-cc60-4519-96f0-68dea8f39c52   
2287   cede56a0-4525-48d4-966a-9bf94a10b4f0   
2329   60226d5b-37b0-400f-aaf6-2427c500c570   
2607   58273161-9375-4610-a3f2-1ee9612e57d8   
2646   75691a58-91dd-474c-aefb-5dbc20bdfa38   
2670   9570e816-5e07-44f0-ac93-7b2bcbc8823b   
2711   8d599e2a-c239-4580-8203-afc2ecc80a3f   
2811   e426fd68-1e66-4c61-a123-9352a9165ba1   
3195   3e909b75-4973-4eb0-81cc-5b19222d8028   
3353   eb02ff84-86cb-4732-9f35-1708ca995162   
3378   c8f875ec-2d2e-4536-9ab0-c727e00b21b5   
3558   d5e2f85c-0b3f-462f-884d-dd29378d269f   
3576   44e4e979-32cc-4fc0-94cd-87eac21c9bc8   
4016   f0787fea-1c46-4e8b-9d09-5fe74af671ec   
4045   d242c38f-8de6-401e-9dc1-359c7052c718   
4227   f8a76c44-ab10-4b62-b581-fabf0a2140cb   
4281   565144e8-e993-46d5-a89e-a536fbfff4df   
4423   427c6e8c-9b93-4646-97ff-08752f0198ce   
4600   9706f75c-0e09-4152-9575-06e8dd12bf05   
5074   66f01c50-0b72-4a24-a36e-15baeb015bdd   
5110   7f2d8cf5-6a4d-475a-a730-77752993cdc4   
5120   7e674699-c8ae-4d10-89a2-7e96de919c39   
5178   74cdcd02-e204-4dca-b521-cb23161a30f7   
5572   d1dc3a3d-0e7a-4e30-aec6-041181d94aba   
5646   bccfdd9c-b7ab-4034-81fa-0bd99480771c   
5797   1ea8e240-a80d-405a-a477-38236a14db25   
5908   12639697-5d1a-4e43-9221-636ddfdfd29b   
6130   83201002-da8d-4164-84fc-4be7bb1c361d   
6168   eda6cade-80a6-4f2a-93a3-321b14eec91f   
6423   b2ca2d63-b38b-4610-a1db-e0a3fbbd5950   
6938   5cdc25e9-e465-4e41-bc3f-d8393c16e228   
7239   f8952f48-41d8-4a7b-983d-6f48440b0163   
7295   db48c240-4d46-4f9d-a7ae-faa0ebcfd81e   
7386   07522178-1724-4266-aa22-72167e7b1d11   
7730   47569e6d-b183-4e16-988f-4b3613a8f1f4   
7943   15cb5925-efec-4a4a-84c8-ca6bf454606c   
8226   109826fe-c237-4d2d-a59f-8118ba0803d3   
8382   319ab68f-ced9-4cb2-b1a7-eaeb4aca5b6a   
8384   3213054d-e76d-4d79-a80b-ad35d525e8ab   
8423   1fdb69ab-6b96-421b-b2fe-d528d1b14ca7   
8439   dade3d5a-fae2-4d65-abb5-538d481f2c03   
8489   60a0b9ee-efdf-4078-a869-5de0793335ee   
8738   81f301b8-cd3b-46f9-a914-c54acff41754   
8891   9ac78991-af8c-425d-a3b7-ca5151eb4671   
9069   e5bae579-1be3-485e-b902-64d37252d25b   
9167   03b57dfe-d567-481e-a93d-a50ca97f1e8b   
9291   77d73e3b-ca8e-428e-878a-1906703f6bcd   
10049  54124250-6f73-4e6d-8528-38b120f36237   
10119  a11a20a7-726d-44bb-913b-341018e3e1f0   
10170  84c3ce83-b6e7-4f60-903f-990aade7f016   
10260  b1a06955-bf27-46df-b8e2-504130648ab1   
10268  7ef86fd6-4f95-4726-b344-682253268326   
10520  32db6239-0d97-494c-a6ce-77cb98a7ee1b   
10529  f3fc087e-01bc-4f7d-ba20-cef71523cdb6   
10582  8f3163ad-f499-473c-9ee9-d624e6ffa77d   
10595  45e407ff-4030-4aad-9531-b1632385dbd7   
11128  42f6a16f-aeec-4635-a235-f77bdf7e6f8a   
11143  e1450514-b620-497d-bacf-fc1ca4cdde97   
11156  53de6245-b637-4e6b-8c2a-fa493620f073   
11309  ac04af5b-632a-4cf6-8507-661b51a7fda2   
11873  41e32c93-9c47-4fe4-9421-9b6c4cf0080c   
12848  6a6461e6-5e0b-4ab1-877e-4a2d5f7b8101   
13156  18de9245-230e-4c5a-8ee3-791f6a0d2142   
13706  8e3541d7-77a3-4a23-8c11-288e05b24459   
13923  e63ee24d-eadd-40d9-b4ab-ef898ad6d6ce   
14259  76f0c15e-8ac9-4dfe-8e89-6d575d3bd633   
14743  98041532-2e84-42e8-85fd-26a7cba77ce4   
14770  02da8794-d87e-4970-8d6a-624f2e0df1a4   
14821  332ce6a5-c3b8-4af4-bd45-7f9ed4855c04   

                                 id_cliente         nombre    apellido  \
557    e13ac13a-2b7f-4fb6-a98f-5cd417859a23       Adelardo      Bernad   
583    cba30122-d7aa-4bca-8b57-70eb0007ff95         Martín    Villalba   
592

## Columna id_hotel

In [29]:
# Tenemos el id del hotel como tipo int, pero nos interesa modificarlo a tipo str/object
data["id_hotel"].dtypes

dtype('int64')

In [30]:
# Realizamos la modificación del tipo de dato
data["id_hotel"] = data["id_hotel"].astype(str)

In [31]:
# Comprobamos que se ha realizado la modificación correctamente
info_df(data)

Tipo_dato  numero_registros  Numero_nulos  %_nulos
id_reserva               object             15098             0     0.00
id_cliente               object             15098             0     0.00
nombre                   object             15098             0     0.00
apellido                 object             15098             0     0.00
mail                     object             15098             0     0.00
competencia                bool             15098             0     0.00
fecha_reserva    datetime64[ns]              9926          5172    34.26
inicio_estancia  datetime64[ns]             15023            75     0.50
final_estancia   datetime64[ns]             15023            75     0.50
id_hotel                 object             15098             0     0.00
precio_noche            float64              9874          5224    34.60
nombre_hotel             object             15098             0     0.00
estrellas               float64              9926          5172    34.26
ciudad                   object             15098             0     0.00
col_prueba               object             15098             0     0.00

In [32]:
arch_limpio = data.to_pickle("../data/reservas_hoteles_limpio.pkl")
arch_limpio